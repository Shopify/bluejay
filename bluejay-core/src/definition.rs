mod arguments_definition;
mod directive_definition;
mod enum_type_definition;
mod enum_value_definition;
mod enum_value_definitions;
mod field_definition;
mod fields_definition;
mod has_directives;
mod input_fields_definition;
mod input_object_type_definition;
mod input_type;
mod input_value_definition;
mod interface_implementation;
mod interface_implementations;
mod interface_type_definition;
mod object_type_definition;
mod output_type;
mod scalar_type_definition;
mod schema_definition;
mod type_definition;
mod union_member_type;
mod union_member_types;
mod union_type_definition;

pub use arguments_definition::ArgumentsDefinition;
pub use directive_definition::{DirectiveDefinition, DirectiveLocation};
pub use enum_type_definition::EnumTypeDefinition;
pub use enum_value_definition::EnumValueDefinition;
pub use enum_value_definitions::EnumValueDefinitions;
pub use field_definition::FieldDefinition;
pub use fields_definition::FieldsDefinition;
pub use has_directives::HasDirectives;
pub use input_fields_definition::InputFieldsDefinition;
pub use input_object_type_definition::InputObjectTypeDefinition;
pub use input_type::{BaseInputTypeReference, InputType, InputTypeReference};
pub use input_value_definition::InputValueDefinition;
pub use interface_implementation::InterfaceImplementation;
pub use interface_implementations::InterfaceImplementations;
pub use interface_type_definition::InterfaceTypeDefinition;
pub use object_type_definition::ObjectTypeDefinition;
pub use output_type::{BaseOutputTypeReference, OutputType, OutputTypeReference};
pub use scalar_type_definition::ScalarTypeDefinition;
pub use schema_definition::SchemaDefinition;
pub use type_definition::{TypeDefinition, TypeDefinitionReference};
pub use union_member_type::UnionMemberType;
pub use union_member_types::UnionMemberTypes;
pub use union_type_definition::UnionTypeDefinition;

pub mod prelude {
    pub use super::{
        ArgumentsDefinition as _, DirectiveDefinition as _, EnumTypeDefinition as _,
        EnumValueDefinition as _, EnumValueDefinitions as _, FieldDefinition as _,
        FieldsDefinition as _, HasDirectives as _, InputFieldsDefinition as _,
        InputObjectTypeDefinition as _, InputType as _, InputValueDefinition as _,
        InterfaceImplementation as _, InterfaceImplementations as _, InterfaceTypeDefinition as _,
        ObjectTypeDefinition as _, OutputType as _, ScalarTypeDefinition as _,
        SchemaDefinition as _, TypeDefinition as _, UnionMemberType as _, UnionMemberTypes as _,
        UnionTypeDefinition as _,
    };
}
