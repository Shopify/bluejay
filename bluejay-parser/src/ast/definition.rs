mod arguments_definition;
mod context;
mod custom_scalar_type_definition;
mod definition_document;
mod directive_definition;
mod enum_type_definition;
mod enum_value_definition;
mod enum_value_definitions;
mod explicit_schema_definition;
mod field_definition;
mod fields_definition;
mod input_fields_definition;
mod input_object_type_definition;
mod input_type;
mod input_value_definition;
mod interface_implementation;
mod interface_implementations;
mod interface_type_definition;
mod object_type_definition;
mod output_type_reference;
mod schema_definition;
mod type_definition_reference;
mod union_member_type;
mod union_member_types;
mod union_type_definition;

pub use arguments_definition::ArgumentsDefinition;
pub use context::{Context, DefaultContext};
pub use custom_scalar_type_definition::CustomScalarTypeDefinition;
pub use definition_document::DefinitionDocument;
pub use directive_definition::DirectiveDefinition;
pub use enum_type_definition::EnumTypeDefinition;
pub use enum_value_definition::EnumValueDefinition;
pub use enum_value_definitions::EnumValueDefinitions;
pub use explicit_schema_definition::{ExplicitSchemaDefinition, RootOperationTypeDefinition};
pub use field_definition::FieldDefinition;
pub use fields_definition::FieldsDefinition;
pub use input_fields_definition::InputFieldsDefinition;
pub use input_object_type_definition::InputObjectTypeDefinition;
pub use input_type::{BaseInputType, InputType};
pub use input_value_definition::InputValueDefinition;
pub use interface_implementation::InterfaceImplementation;
pub use interface_implementations::InterfaceImplementations;
pub use interface_type_definition::InterfaceTypeDefinition;
pub use object_type_definition::ObjectTypeDefinition;
pub use output_type_reference::{BaseOutputType, OutputTypeReference};
pub use schema_definition::SchemaDefinition;
pub use type_definition_reference::TypeDefinitionReference;
pub use union_member_type::UnionMemberType;
pub use union_member_types::UnionMemberTypes;
pub use union_type_definition::UnionTypeDefinition;
