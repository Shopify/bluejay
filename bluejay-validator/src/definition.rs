mod error;
mod rule;
mod rules;
mod validator;
mod visitor;

pub use error::Error;
pub use rule::Rule;
pub use rules::BuiltinRules;
pub use validator::{BuiltinRulesValidator, Validator};
pub use visitor::Visitor;
