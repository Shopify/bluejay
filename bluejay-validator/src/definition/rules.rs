mod enum_value_definition_uniqueness;
mod input_field_definition_uniqueness;

pub use enum_value_definition_uniqueness::EnumValueDefinitionUniqueness;
pub use input_field_definition_uniqueness::InputFieldDefinitionUniqueness;

#[macro_export]
macro_rules! combine_definition_rules {
    ( $name:ty, $err:ty, [$( $rule:ty ),* $(,)?] $(,)? ) => {
        paste::paste! {
            pub struct $name<'a, S: bluejay_core::definition::SchemaDefinition> {
                $([<$rule:snake>]: $rule<'a, S>,)*
            }

            impl<'a, S: bluejay_core::definition::SchemaDefinition + 'a> $crate::definition::Rule<'a, S> for $name<'a, S> {
                type Error = $err<'a, S>;

                fn new(schema_definition: &'a S) -> Self {
                    Self {
                        $([<$rule:snake>]: $rule::new(schema_definition),)*
                    }
                }
            }

            impl<'a, S: bluejay_core::definition::SchemaDefinition + 'a> IntoIterator for $name<'a, S> {
                type Item = $err<'a, S>;
                type IntoIter = $crate::chain_types!($(std::iter::Map<<$rule<'a, S> as IntoIterator>::IntoIter, fn(<$rule<'a, S> as $crate::definition::Rule<'a, S>>::Error) -> $err<'a, S>>),*);

                fn into_iter(self) -> Self::IntoIter {
                    $crate::chain_iters!($(self.[<$rule:snake>].into_iter().map(Into::into as fn(<$rule<'a, S> as $crate::definition::Rule<'a, S>>::Error) -> $err<'a, S>)),*)
                }
            }

            impl<'a, S: bluejay_core::definition::SchemaDefinition> $crate::definition::Visitor<'a, S> for $name<'a, S> {
                fn visit_input_object_type_definition(&mut self, input_object_type_definition: &'a S::InputObjectTypeDefinition) {
                    $(self.[<$rule:snake>].visit_input_object_type_definition(input_object_type_definition);)*
                }

                fn visit_enum_type_definition(&mut self, enum_type_definition: &'a S::EnumTypeDefinition) {
                    $(self.[<$rule:snake>].visit_enum_type_definition(enum_type_definition);)*
                }
            }
        }
    };
}

combine_definition_rules!(
    BuiltinRules,
    crate::definition::Error,
    [
        EnumValueDefinitionUniqueness,
        InputFieldDefinitionUniqueness,
    ],
);
