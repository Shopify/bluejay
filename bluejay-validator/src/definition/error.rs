use bluejay_core::definition::SchemaDefinition;

#[cfg(feature = "parser-integration")]
use bluejay_parser::{
    ast::definition::SchemaDefinition as ParserSchemaDefinition,
    error::{Annotation, Error as ParserError},
    HasSpan,
};

pub enum Error<'a, S: SchemaDefinition> {
    NonUniqueInputValueDefinitionNames {
        name: &'a str,
        input_value_definitions: Vec<&'a S::InputValueDefinition>,
    },
    NonUniqueEnumValueDefinitionNames {
        name: &'a str,
        enum_value_definitions: Vec<&'a S::EnumValueDefinition>,
    },
}

#[cfg(feature = "parser-integration")]
impl<'a> From<Error<'a, ParserSchemaDefinition<'a>>> for ParserError {
    fn from(value: Error<'a, ParserSchemaDefinition<'a>>) -> Self {
        match value {
            Error::NonUniqueInputValueDefinitionNames {
                name,
                input_value_definitions,
            } => Self::new(
                format!("Multiple input value definitions named `{name}`"),
                None,
                input_value_definitions
                    .into_iter()
                    .map(|ivd| {
                        Annotation::new(
                            format!("Input value definition with name `{name}`"),
                            ivd.name_token().span().clone(),
                        )
                    })
                    .collect(),
            ),
            Error::NonUniqueEnumValueDefinitionNames {
                name,
                enum_value_definitions,
            } => Self::new(
                format!("Multiple enum value definitions named `{name}`"),
                None,
                enum_value_definitions
                    .into_iter()
                    .map(|evd| {
                        Annotation::new(
                            format!("Enum value definition with name `{name}`"),
                            evd.name_token().span().clone(),
                        )
                    })
                    .collect(),
            ),
        }
    }
}
