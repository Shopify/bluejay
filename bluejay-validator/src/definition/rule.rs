use crate::definition::Visitor;
use bluejay_core::definition::SchemaDefinition;

pub trait Rule<'a, S: SchemaDefinition>: Visitor<'a, S> + IntoIterator<Item = Self::Error> {
    type Error;

    fn new(schema_definition: &'a S) -> Self;
}
