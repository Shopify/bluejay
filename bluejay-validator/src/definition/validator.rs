use crate::definition::{BuiltinRules, Rule};
use bluejay_core::definition::{SchemaDefinition, TypeDefinitionReference};

pub struct Validator<'a, S: SchemaDefinition, R: Rule<'a, S>> {
    schema_definition: &'a S,
    rule: R,
}

pub type BuiltinRulesValidator<'a, S> = Validator<'a, S, BuiltinRules<'a, S>>;

impl<'a, S: SchemaDefinition, R: Rule<'a, S>> Validator<'a, S, R> {
    fn new(schema_definition: &'a S) -> Self {
        Self {
            schema_definition,
            rule: Rule::new(schema_definition),
        }
    }

    fn visit(&mut self) {
        self.schema_definition.type_definitions().for_each(
            |type_definition| match type_definition {
                TypeDefinitionReference::InputObject(iotd) => {
                    self.visit_input_object_type_definition(iotd)
                }
                TypeDefinitionReference::Enum(etd) => self.visit_enum_type_definition(etd),
                _ => {}
            },
        )
    }

    fn visit_input_object_type_definition(
        &mut self,
        input_object_type_definition: &'a S::InputObjectTypeDefinition,
    ) {
        self.rule
            .visit_input_object_type_definition(input_object_type_definition);
    }

    fn visit_enum_type_definition(&mut self, enum_type_definition: &'a S::EnumTypeDefinition) {
        self.rule.visit_enum_type_definition(enum_type_definition);
    }

    pub fn validate(schema_definition: &'a S) -> <Self as IntoIterator>::IntoIter {
        let mut instance = Self::new(schema_definition);
        instance.visit();
        instance.into_iter()
    }
}

impl<'a, S: SchemaDefinition, R: Rule<'a, S>> IntoIterator for Validator<'a, S, R> {
    type Item = R::Error;
    type IntoIter = <R as IntoIterator>::IntoIter;

    fn into_iter(self) -> Self::IntoIter {
        self.rule.into_iter()
    }
}
