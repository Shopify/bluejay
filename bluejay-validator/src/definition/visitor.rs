use bluejay_core::definition::SchemaDefinition;

pub trait Visitor<'a, S: SchemaDefinition> {
    fn visit_input_object_type_definition(
        &mut self,
        _input_object_type_definition: &'a S::InputObjectTypeDefinition,
    ) {
    }

    fn visit_enum_type_definition(&mut self, _enum_type_definition: &'a S::EnumTypeDefinition) {}
}
