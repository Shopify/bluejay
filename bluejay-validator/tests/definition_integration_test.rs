use bluejay_parser::{
    ast::{
        definition::{DefinitionDocument, SchemaDefinition},
        Parse,
    },
    Error,
};
use bluejay_validator::definition::BuiltinRulesValidator;

#[test]
fn test_error() {
    insta::glob!("test_data/definition/error/*.graphql", |path| {
        let input = std::fs::read_to_string(path).unwrap();
        let definition_document: DefinitionDocument = DefinitionDocument::parse(&input)
            .unwrap_or_else(|_| panic!("Schema `{}` had parse errors", path.display()));
        let schema_definition = SchemaDefinition::try_from(&definition_document)
            .unwrap_or_else(|_| panic!("Schema `{}` had coercion errors", path.display()));

        let errors: Vec<_> = BuiltinRulesValidator::validate(&schema_definition).collect();

        let formatted_errors =
            Error::format_errors(&input, path.file_name().and_then(|f| f.to_str()), errors);
        insta::assert_snapshot!(formatted_errors);
    });
}

#[test]
fn test_valid() {
    insta::glob!("test_data/definition/valid/*.graphql", |path| {
        let input = std::fs::read_to_string(path).unwrap();
        let definition_document: DefinitionDocument = DefinitionDocument::parse(&input)
            .unwrap_or_else(|_| panic!("Schema `{}` had parse errors", path.display()));
        let schema_definition = SchemaDefinition::try_from(&definition_document)
            .unwrap_or_else(|_| panic!("Schema `{}` had coercion errors", path.display()));

        let errors: Vec<_> = BuiltinRulesValidator::validate(&schema_definition).collect();

        assert!(
            errors.is_empty(),
            "Schema `{}` had validation errors:\n{}",
            path.display(),
            Error::format_errors(&input, path.file_name().and_then(|f| f.to_str()), errors),
        )
    });
}
