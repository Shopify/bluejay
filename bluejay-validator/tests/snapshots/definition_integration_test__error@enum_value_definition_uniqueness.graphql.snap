---
source: bluejay-validator/tests/definition_integration_test.rs
expression: formatted_errors
input_file: bluejay-validator/tests/test_data/definition/error/enum_value_definition_uniqueness.graphql
---
Error: Multiple enum value definitions named `VARIANT_1`
   ╭─[ enum_value_definition_uniqueness.graphql:1:1 ]
   │
 2 │   VARIANT_1
   │   ────┬────  
   │       ╰────── Enum value definition with name `VARIANT_1`
   │ 
 4 │   VARIANT_1
   │   ────┬────  
   │       ╰────── Enum value definition with name `VARIANT_1`
───╯
